client/state/user-settings/helpers.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * Set a (nested) value in an object. Will create keys if they're not existent.
+ *
+ * @param {object} obj Object to operate on
+ * @param {string | Array} path Key to set. Can be an array to represent nested values
+ * @param {string} value Value which key should be set to
+ */
+export function setValue( obj, path, value ) {
+	if ( ! Array.isArray( path ) ) {
+		path = [ path ];
+	}
+	function doSet( o = {}, i ) {
+		return {
+			...o,
+			[ path[ i ] ]: i === path.length - 1 ? value : doSet( o[ path[ i ] ], i + 1 ),
```

Comment:
If we passed `path` as an argument to `doSet` could we prevent from redeclaring `doSet` on every `setValue` call? Same question for the `doRemove` below.

Reply:
Yes, if `doSet` gets all the data it needs from parameters instead of a closure, it could be a standalone `doSet( obj, value, path, pathIndex )` function. It's the same with `doRemove`.

Reply:
Is there a real benefit to extracting `doSet` and `doRemove`? I actually quite like that it's encapsulated inside as closure because those functions are not intended to be used anywhere else.

Reply:
In my mind, the benefit is purely theoretical and won't have any impact on performance. Feel free to leave it as-is.

Reply:
Theoretically the code with the extracted functions should be faster because you don't need to re-create an instance of `doSet` and `doRemove` on every call. But in practice the JS engine might optimize it in such a way that it doesn't matter.

Reply:
This.